### DISCUSSION TITLE
Create read and write functionality from Modbus to OPCUA using python

### DISCUSSION THREAD

**Discussion by vidyarao123** [FreeOpcUa/python-opcua#1236] [Q&A]

**Comment by AndreasHeine** [comment]
looks like you read and write to the same modbus register!?

**Reply by vidyarao123** [reply]
Yes

**Reply by vidyarao123** [reply]
I am considering the holding register for the same purpose ( Read and Write )

**Reply by AndreasHeine** [reply]
reading and writing the same register from two places makes things complicated and unpredictable... you should consider define directional register areas like: reg0-15 (Modbus -> OPC UA) and reg 16-32 (OPC UA -> Modbus) to avoide overwriting each other (if i remember right ther was a memory map for modbus wich has like readonly regs as "Inputs" and read/write regs as "Outputs")

**Comment by swamper123** [comment]
1. Normally we would be just responsible for the OPC UA part. Unless you don't read the same thing you have wrote into the Node (and you don't do crazy server magic or manipulating the value otherwise, before reading it again), it wouldn't be the scope to discuss here. Questions about your ModbusClient should be discussed in your Modbus Repo.
2. See the answer from @AndreasHeine. You may want to change sth. (value or register) to recognize a change?
3. Another question about Modbus here cost a virtual Pizza for all Contributers. :^)

**Reply by vidyarao123** [reply]
Hey,

I wanted to know from the OPC UA part only. Is there a way to monitor the event change from OPC Server when the value is written from OPC CLient. 

Sorry for confusing you earlier.

**Reply by AndreasHeine** [reply]
you can use the subscription service of opc ua to get valuechanges! so the opc ua client gets each datachange event from the opcua server.

**Reply by vidyarao123** [reply]
The subscription service is for client. But my python code is acting as OPC UA server. And UA Expert will be my opc client. Will the subscription server still work in this case ?

**Reply by AndreasHeine** [reply]
subscriptions work also on serverside!

**Reply by vidyarao123** [reply]
Wonderful..  I will try and get back to you on this. Thank you so much;)

**Comment by zerox1212** [comment]
I have an outdated example repo that was on a working system.

https://github.com/zerox1212/python-opcua-modbus-driver
